### PR TITLE
sdk automation, support breaking change detection

### PR DESCRIPTION
**Pull request by XiaofeiCao**
# Description

- fix https://github.com/Azure/azure-sdk-tools/issues/8477
- supported
  - mgmt from TypeSpec, test: https://github.com/test-repo-billy/azure-rest-api-specs/pull/3381
  - mgmt from Swagger, test: https://github.com/test-repo-billy/azure-rest-api-specs/pull/3382
  - dpg from TypeSpec, test: https://github.com/test-repo-billy/azure-rest-api-specs/pull/3383
  - dpg from Swagger(Tested locally)

For SDK Automation, breaking change is compared with last GA version(new feature, this is about this PR).
For mgmt SDK generation, breaking change is compared with last released version(existing feature, not related to this PR).

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: eng/mgmt/automation/generate.py [hunk 1]
```diff
@@ -244,12 +254,18 @@ def sdk_automation_typespec_project(tsp_project: str, config: dict) -> dict:
...
         # TODO (weidxu): move to typespec-java
         if require_sdk_integration:
-            set_or_default_version(sdk_root, GROUP_ID, module)
             update_service_ci_and_pom(sdk_root, service, GROUP_ID, module)
             update_root_pom(sdk_root, service)
 
+        stable_version, current_version = set_or_increase_version(sdk_root, GROUP_ID, module)
+        update_parameters(None)
+        output_folder = OUTPUT_FOLDER_FORMAT.format(service)
+        update_version(sdk_root, output_folder)
```
Comment (XiaofeiCao) [inline eng/mgmt/automation/generate.py:263]: update version for mgmt TypeSpec, to align with mgmt Swagger in `generate`

File: eng/mgmt/automation/generate_utils.py [hunk 2]
```diff
@@ -189,33 +189,20 @@ def update_changelog(changelog_file, changelog):
...
-        version_pattern = r"\d+\.\d+\.\d+-beta\.(\d+)?"
-        beta_version_int = int(re.match(version_pattern, current_version).group(1))
-        if beta_version_int > 1:
-            previous_beta_version_int = beta_version_int - 1
-            previous_beta_version = current_version.replace(
-                "-beta." + str(beta_version_int),
-                "-beta." + str(previous_beta_version_int),
-            )
-            stable_version = previous_beta_version
+        return False, ""
```
Comment (XiaofeiCao) [inline eng/mgmt/automation/generate_utils.py:195]: moved this part to `utils.get_latest_release_version`

File: eng/mgmt/automation/generate.py [hunk 3]
```diff
@@ -178,7 +180,10 @@ def sdk_automation_autorest(config: dict) -> List[dict]:
                 tag=tag,
             )
             if succeeded:
-                compile_arm_package(sdk_root, module)
+                compile_succeeded = compile_arm_package(sdk_root, module)
+                if compile_succeeded:
+                    stable_version = get_latest_ga_version(GROUP_ID, module, stable_version)
+                    breaking, changelog = compare_with_maven_package(sdk_root, GROUP_ID, service, stable_version, current_version, module)
```
Comment (XiaofeiCao) [inline eng/mgmt/automation/generate.py:186]: mgmt Swagger

File: eng/mgmt/automation/generate.py [hunk 4]
```diff
@@ -244,12 +254,18 @@ def sdk_automation_typespec_project(tsp_project: str, config: dict) -> dict:
...
 
+        stable_version, current_version = set_or_increase_version(sdk_root, GROUP_ID, module)
+        update_parameters(None)
+        output_folder = OUTPUT_FOLDER_FORMAT.format(service)
+        update_version(sdk_root, output_folder)
+
         # compile
         succeeded = compile_arm_package(sdk_root, module)
+        if succeeded:
+            breaking, changelog = compare_with_maven_package(sdk_root, GROUP_ID, service, get_latest_ga_version(GROUP_ID, module, stable_version), current_version, module)
```
Comment (XiaofeiCao) [inline eng/mgmt/automation/generate.py:268]: mgmt TypeSpec

File: eng/mgmt/automation/generate.py [hunk 5]
```diff
@@ -343,7 +363,8 @@ def main():
     if succeeded:
         succeeded = compile_arm_package(sdk_root, module)
         if succeeded:
-            compare_with_maven_package(sdk_root, service, stable_version, current_version, module)
+            latest_release_version = get_latest_release_version(current_version, stable_version)
+            compare_with_maven_package(sdk_root, GROUP_ID, service, latest_release_version, current_version, module)
```
Comment (XiaofeiCao) [inline eng/mgmt/automation/generate.py:367]: Just a refactor. Calculate last release version outside `compare_with_maven_package`.

File: eng/mgmt/automation/generate_data.py [hunk 6]
```diff
@@ -46,28 +50,31 @@ def sdk_automation_typespec_project(tsp_project: str, config: dict) -> dict:
...
-            set_or_default_version(sdk_root, GROUP_ID, module)
             update_service_ci_and_pom(sdk_root, service, GROUP_ID, module)
             update_root_pom(sdk_root, service)
 
         # compile
         succeeded = compile_package(sdk_root, GROUP_ID, module)
-
-        if not succeeded:
+        if succeeded:
+            breaking, changelog = compare_with_maven_package(sdk_root, GROUP_ID, service, get_latest_ga_version(GROUP_ID, module, stable_version), current_version, module)
```
Comment (XiaofeiCao) [inline eng/mgmt/automation/generate_data.py:61]: dpg TypeSpec

File: eng/mgmt/automation/generate_data.py [hunk 7]
```diff
@@ -224,8 +235,14 @@ def sdk_automation_readme(readme_file_abspath: str, packages: List[dict], sdk_ro
...
 
+        breaking = False
+        changelog = ""
+
         if succeeded:
-            compile_package(sdk_root, GROUP_ID, module)
+            stable_version, current_version = set_or_default_version(sdk_root, GROUP_ID, module)
+            succeeded = compile_package(sdk_root, GROUP_ID, module)
+            if succeeded:
+                breaking, changelog = compare_with_maven_package(sdk_root, GROUP_ID, service, get_latest_ga_version(GROUP_ID, module, stable_version), current_version, module)
```
Comment (XiaofeiCao) [inline eng/mgmt/automation/generate_data.py:245]: dpg Swagger

File: eng/mgmt/automation/utils.py [hunk 8]
```diff
@@ -350,3 +353,83 @@ def set_or_increase_version(
...
+            )
+            previous_version = previous_beta_version
+    return previous_version
+
+
+def get_latest_ga_version(group_id: str, module: str, previous_version: str) -> str:
+    """
+    If previous_version is GA, return previous_version,
+    Otherwise, return previous GA version before previous beta.
+    previous_version itself if no previous GA version.
```
Comment (weidongxu-microsoft) [inline eng/mgmt/automation/utils.py:378]: Why return non-GA version, if the method is get_latest_ga_version?

File: eng/mgmt/automation/generate.py [hunk 9]
```diff
@@ -343,7 +363,8 @@ def main():
     if succeeded:
         succeeded = compile_arm_package(sdk_root, module)
         if succeeded:
-            compare_with_maven_package(sdk_root, service, stable_version, current_version, module)
+            latest_release_version = get_latest_release_version(current_version, stable_version)
```
Comment (weidongxu-microsoft) [inline eng/mgmt/automation/generate.py:366]: Why call this on `stable_version`?

Comment (XiaofeiCao) [reply eng/mgmt/automation/generate.py]: Right, thanks... Didn't test codegen.
Fixed and test passed in local generation.

File: eng/mgmt/automation/utils.py (hunk 8 above)
Comment (XiaofeiCao) [reply eng/mgmt/automation/utils.py:378]: Good point. This was for later check in `compare_with_maven_package`. 
Though a second thought, this should only return GA version, and None if no GA version found. Later check should check for None also.

Comment (XiaofeiCao) [reply eng/mgmt/automation/utils.py]: Fixed and tested locally.

**Comment by XiaofeiCao** [comment]
@raych1 Would you help take a look at the GenerateApiView error in pipeline?
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3918820&view=logs&j=a8a7a537-82b0-583c-7971-bac70b9822ca&t=61443e36-1bb1-529f-ff2d-19827770e639

It's looking for non-exist artifact `azure-resourcemanager-scvmm-1.1.0-beta.1-sources.json.tgz`, which should be `azure-resourcemanager-scvmm-1.1.0-beta.1-sources.jar`. 
Not sure if this is related to the change made...

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by raych1** [comment]
> For SDK Automation, breaking change is compared with last GA version(new feature). For mgmt SDK generation, breaking change is compared with last released version(existing feature).

@XiaofeiCao 
can you explain why Java uses different base versions for above two scenarios?

Regarding APIView failures, I noticed that it fails for Java in other test PRs too. But API view was created successfully for other languages. We shall take a further look. 
for example, [this PR](https://github.com/test-repo-billy/azure-rest-api-specs/pull/3389)

**Comment by weidongxu-microsoft** [comment]
@raych1 

The "mgmt SDK generation" part is for generating SDK for release. No changelog nor breaking change flag would sent to you or your pipeline. Here we uses our existing logic (mostly, GA compare to GA, preview compare to a prior version) for SDK changelog.

Only the "SDK Automation" relates to you / PR.

**Comment by XiaofeiCao** [comment]
> @XiaofeiCao can you explain why Java uses different base versions for above two scenarios?

We didn't touch existing codegen logic, like Weidong explained above. This PR's only about SDK automation. 
I've updated the description. 

> Regarding APIView failures, I noticed that it fails for Java in other test PRs too. But API view was created successfully for other languages. We shall take a further look. for example, [this PR](https://github.com/test-repo-billy/azure-rest-api-specs/pull/3389)

Thanks. Java's error is strange, since the output apiViewArtifact looks OK to me: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3923824&view=logs&j=a8a7a537-82b0-583c-7971-bac70b9822ca&t=37e3947b-3cfb-5d36-86ba-0e22bb7dbc33&l=1484

**Comment by raych1** [comment]
@XiaofeiCao ,

APIView has been generated successfully in this PR for Java, the failure in this PR should not be related to your changes.
https://github.com/test-repo-billy/azure-rest-api-specs/pull/3390#issuecomment-2202249969

**Review by raych1 — APPROVED** [review]

**Comment by XiaofeiCao** [comment]
> @XiaofeiCao ,
> 
> APIView has been generated successfully in this PR for Java, the failure in this PR should not be related to your changes. [test-repo-billy/azure-rest-api-specs#3390 (comment)](https://github.com/test-repo-billy/azure-rest-api-specs/pull/3390#issuecomment-2202249969)

Thanks @raych1 !

**Pull request merged**